nixos/modules/services/network-filesystems/ipfs.nix

```diff
@@ -9,7 +9,10 @@ let
 
   ipfsFlags = ''${if cfg.autoMigrate then "--migrate" else ""} ${if cfg.enableGC then "--enable-gc" else ""} ${toString cfg.extraFlags}'';
 
-  pathEnv = { IPFS_PATH = cfg.dataDir; };
+  # Before Version 17.09, ipfs would always use "/var/lib/ipfs/.ipfs" as it's dataDir
+  defaultDataDir = if versionAtLeast config.system.stateVersion "17.09" then
+    "/var/lib/ipfs" else
+    "/var/lib/ipfs/.ipfs";
 
```

Comment:
But what might be a problem is that even if the user had set `dataDir` before, it was actually `dataDir/.ipfs` which was used, I'm actually not sure what to do about this. I think such a script wouldn't be too bad, could be run once at build time. Even if there aren't many users, I think it would be a shame to have it break on the user when upgrading.